### DISCUSSION TITLE
Add accessible labels to form fields

### DISCUSSION THREAD

**Discussion by ubercj** [directus/directus#21298] [Feature Requests]
### Summary

I am a developer for a large client who is using Directus. We love the features of the base platform and the flexibility of the extensions SDK, but we are concerned about the accessibility of certain parts of the app. There are several areas where I think massive improvements could be made with relatively small investment.

Forms (for creating Items, Collections, etc.) are Directus' bread and butter and so I think it's a great place to start. Currently, many form controls don't appear to be tied to a label. Addressing this would have a massive impact on accessibility site-wide.

### Basic Example

I ran a quick Lighthouse audit of a form to create a new Item:
<img width="973" alt="Screenshot 2024-01-25 at 1 04 41 PM" src="https://github.com/directus/directus/assets/73803823/c62b4894-7442-4820-90fd-04b646b29b27">

Here are screenshots of both a basic text input with its visual label, and how that appears in the accessibility tree in Chrome dev tools:
<img width="881" alt="Screenshot 2024-02-03 at 2 56 00 PM" src="https://github.com/directus/directus/assets/73803823/52551091-cf1d-40c5-b9fa-75c6d51f19ff">
<img width="793" alt="Screenshot 2024-02-03 at 2 56 13 PM" src="https://github.com/directus/directus/assets/73803823/b6531059-fc08-4690-a872-681dba911dc5">

That `textbox ""` indicates no label for the input. It should be `textbox "Name"`, to match the visual label.

### Motivation

Setting aside that improving the accessibility of the site is a good thing to do in its own right, I think this could be a factor in companies who could use Directus abandoning it in favor of other options with better out-of-the-box accessibility. I'm thinking companies that must meet legal requirements for accessibility like schools, government agencies/contractors, etc.

### Detailed Design

My first thought was that you could just add the `aria-label` attribute on the `form-field-interface`. You could use `field.name` since that is the value being used in `form-field-label`.

This probably won't work since you've got some complex interfaces with multiple layers of elements between the `<component>` tag and the actual form control. This makes it much less straightforward, but it should be possible with just the props already given to the interface component, you've just got to get `field.name` down to the element somehow. An example for text inputs might be something like this.

```ts
// app/src/components/v-form/form-field-interface.vue
<component
  :is="componentName"
  // ...
  :field="field.field"
  :field-data="field"
  // ...
/>
```

```ts
// app/src/interfaces/input/input.vue
const props = withDefaults(
  defineProps<{
  value: string | number | null;
  type?: string;
  clear?: boolean;
  // + field: FormField
  // ...

<v-input
  :autofocus="autofocus"
  :model-value="value"
  // + :field-label="field.name"
  // ...
```

```ts
// app/src/components/v-input.vue
interface Props {
  /** Autofocusses the input on render */
  autofocus?: boolean;
  // + fieldLabel: string;
  // ...

<slot name="input">
  <input
    ref="input"
    v-focus="autofocus"
    v-bind="attributes"
    :placeholder="placeholder ? String(placeholder) : undefined"
    // + :aria-label="fieldLabel"
```

Instead of prop drilling, you could consider using provide/inject. I'm not sure if the fields store could come into play either

### Must Have:
- Accessible names for form controls. This could be narrowed down to do one at a time. Text inputs could be a place to start.

### Should Have:
- NA

### Could Have:
- NA

### Won't Have:
- NA

### Drawbacks

- This process is already fairly intricate with just text inputs. It could be even less straightforward for more complex form controls like datepickers or dropdowns
- I don't know where you are in the process of turning the internal v-components into its own package and whether or not that complicates things

### Alternatives

- You could do something complicated with `aria-hidden` and a [`visually-hidden` class](https://www.tpgi.com/the-anatomy-of-visually-hidden/) to essentially include an entirely separate duplicated form that is accessible to only assistive tech. This might make it possible to avoid messing with the existing setup for fields, but it would require quite a bit of extra JS to wire up state updates for the second hidden form
- You could decide this isn't worth your time. Certain people with disabilities are shut out from using Directus and it could be a dealbreaker for potential clients/customers

### Adoption Strategy

I would recommend starting small and taking a piecemeal approach. Maybe do 1 or 2 interface types in a sprint. If you have an existing workflow for addressing tech debt, this could be included in that

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
